proof_of_concept/components/domain_administrator.py

```diff
@@ -0,0 +1,320 @@
+"""Components that manage local container-based execution."""
+import gzip
+import json
+import logging
+from threading import Lock
+import tempfile
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import docker
+from docker.models.images import Image
+from docker.models.containers import Container
+
+from proof_of_concept.components.asset_store import AssetStore
+from proof_of_concept.definitions.assets import (
+        Asset, ComputeAsset, DataAsset, Metadata)
+from proof_of_concept.definitions.identifier import Identifier
+from proof_of_concept.definitions.interfaces import IDomainAdministrator
+from proof_of_concept.definitions.workflows import Job, WorkflowStep
+from proof_of_concept.rest.client import SiteRestClient
+
+
+OUTPUT_ASSET_ID = Identifier('asset:ns:output_base:ns:site')
+
+
+logger = logging.getLogger(__name__)
+
+
+class PlainDockerDA(IDomainAdministrator):
```

Comment:
I would call this StepRunner or DockerStepRunner or something and not DomainAdministrator, because to me that is what this class does. But that is because I don't know what it means to administrate a domain, so that might be a fancy name for the same thing that I just need to learn...

Reply:
Well, I already have a StepRunner class, so I couldn't use that, and there will be other runners that use Docker to start the containers, but do it in other ways. So I wasn't sure what to call this, and ended up going with the term that the scientists came up with. I think "domain" refers to an administrative domain, e.g. it's the kingdom of the sysadmin/network engineer in charge of the site, and "administrator" means that it controls the things (containers, network connections) in that domain. I'm not super-happy with it, but don't really have a better solution either.

Reply:
Added some documentation to this effect.